File: MultiLanguageSandbox/src/test/deps/ruby/20.rb


# Finds the intersection point of two linear functions.
# This function takes four arguments: slope1, intercept1, slope2, intercept2.
# slope1 and slope2 are the slopes of the two lines, while intercept1 and intercept2 are their y-intercepts.
# The function calculates and returns the coordinates of the intersection point of these two lines, if it exists.
# The coordinates are returned as a two-element array [x, y], rounded to two decimal places.
#
# Example:
# >>> find_intersection(1.0, 0.0, -1.0, 2.0)
# [1.00, 1.00]
# >>> find_intersection(2.0, 3.0, 2.0, -1.0)
# nil # Parallel lines with no intersection


def find_intersection(slope1, intercept1, slope2, intercept2)
  # Check if the lines are parallel (same slope)
  if slope1 == slope2
    return nil  # No intersection or infinite intersections (same line)
  end
  
  # Calculate x-coordinate of intersection
  x = (intercept2 - intercept1) / (slope1 - slope2)
  
  # Calculate y-coordinate using either line equation
  y = slope1 * x + intercept1
  
  # Round to two decimal places and return as array
  [x.round(2), y.round(2)]
end

raise 'Test failed' unless find_intersection(1.0, 0.0, -1.0, 2.0) == [1.00, 1.00]
raise 'Test failed' unless find_intersection(2.0, 3.0, 2.0, -1.0).nil?
raise 'Test failed' unless find_intersection(0.5, 1.0, -0.5, -1.0) == [-2.00, 0.00]
raise 'Test failed' unless find_intersection(3.0, 2.0, -1.0, 4.0) == [0.50, 3.50]
raise 'Test failed' unless find_intersection(2.0, 0.0, 0.0, 2.0) == [1.0, 2.0]
  

puts 'All tests passed!'

File: MultiLanguageSandbox/src/test/deps/rust/20.rs

/*
  Calculates the area of a triangle given the coordinates of its vertices.

  Arguments:
  - x1, y1: The x and y coordinates of the first vertex.
  - x2, y2: The x and y coordinates of the second vertex.
  - x3, y3: The x and y coordinates of the third vertex.

  Returns:
  - The area of the triangle, rounded to two decimal places.

  Examples:
  - triangle_area(0.0, 0.0, 4.0, 0.0, 0.0, 3.0) returns 6.00
*/


fn triangle_area(x1: f32, y1: f32, x2: f32, y2: f32, x3: f32, y3: f32) -> f32 {
    // Calculate the area using the shoelace formula
    let area = 0.5 * ((x2 - x1) * (y3 - y1) - (x3 - x1) * (y2 - y1)).abs();
    
    // Round to two decimal places
    (area * 100.0).round() / 100.0
}

// Example usage:


#[cfg(test)]
mod tests {
    use super::*;
 
    #[test]
    fn main() {
        assert_eq!(triangle_area(0.0, 0.0, 4.0, 0.0, 0.0, 3.0), 6.00);
        assert_eq!(triangle_area(1.0, 2.0, 4.0, 5.0, 6.0, 3.0), 6.00); // Example test case, replace with correct area
        assert_eq!(triangle_area(0.0, 0.0, 7.0, 8.0, 9.0, 10.0), 1.00); 
    }
    

}